Add new Activity: activity_library.xml, activity_search.xml, activity…

KiMate34 opened the pull request:
…_settings.xml

Add new class: LibraryActivity, SearchActivity, SettingsActivity

KiMate34 committed:
Add new Activity: activity_library.xml, activity_search.xml, activity_settings.xml
Add new class: LibraryActivity, SearchActivity, SettingsActivity

faritowich commented on app/src/main/java/com/example/playlistmaker/SettingsActivity.kt:
```diff
@@ -0,0 +1,19 @@
+        val buttonBackMain = findViewById<Button>(R.id.button_back)
+        buttonBackMain.setOnClickListener{
+            val backMainIntent = Intent(this, MainActivity::class.java)
+            startActivity(backMainIntent)
```
Для перехода назад надо использовать не интент, а методы finish() или onBackPressed(), потому что иначе вместо возврата в предыдущий экран создается новый экземпляр MainActivity и при возврате обратно на главный экран и нажатии на кнопку телефона "назад" мы "вернемся" снова на этот экран. На эту тему можно почитать про стек активити. Это будет критично в следующем спринте.

faritowich commented on app/src/main/res/values/dimens.xml:
```diff
@@ -0,0 +1,10 @@
+<resources>
+    <dimen name="textSize">22sp</dimen>
+    <dimen name="main_button_cornerRadius">22dp</dimen>
+    <dimen name="settings_frame_height">61dp</dimen>
```
Рекомендуется делать размеры View и отступы кратными 8 (или 4). Это обеспечивает единообразие дизайна, а также связано с тем, что большинство размеров экрана делятся на 8. Иногда эту рекомендацию приходится нарушать, но лучше по возможности ей следовать.
Также см. https://material.io/design/layout/spacing-methods.html#baseline-grid

Дизайнеры не всегда следуют рекомендациям гугла. У нас задача сделать такой UI, чтобы он соответствовал макетам внешне, не обязательно пиксель в пиксель, тем более что это невозможно, учитывая огромное количество форм и размеров устройств. Так что если мы сделаем и UI как на макетах, и при этом последуем рекомендациям гугла, то будет круто

faritowich commented on app/src/main/res/values/dimens.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <dimen name="textSize">22sp</dimen>
```
Очень общее название у размера шрифта, лучше конкретизировать

faritowich commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,89 @@
+            android:text="@string/app_name"
+            android:textColor="@color/white"
+            android:textSize="22sp" />
+    </FrameLayout>
```
Лишняя вложенность, FrameLayout не нужен

faritowich commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,89 @@
+            app:iconTint="@color/black"
+            app:iconGravity="textStart"
+            app:iconPadding="8dp"
+            app:iconSize="22dp"
```
iconPadding и iconSize - повторяющиеся размеры, стоит вынести в ресурсы

faritowich commented on app/src/main/res/layout/activity_settings.xml:
```diff
@@ -0,0 +1,121 @@
+            android:text="@string/settings"
+            android:textColor="@color/black"
+            android:textSize="@dimen/textSize" />
+    </FrameLayout>
```
Не очень хорошо TextView расположен с этим layout_marginEnd + layout_gravity. Тут напрашивается просто горизонтальный LinearLayout вместо FrameLayout, тогда не придется эти лишние приседания делать

KiMate34 replied:
Да, я с вами полностью согласен, так как при повороте экрана эта View просто уезжала в центр) Я сделал по рекомендации и получилось действительно правильнее, но marginTop всё же добавил, так как TextView относительно Button находился не по центру. Может есть какой-то другой способ сделать их на одном уровне и при этом оставить код эстетически красивым?

faritowich replied:
Сходу три варианта, можно попробовать и решить, какой лучше:
1) Можно использовать TextView и задавать иконку при помощи атрибута drawableStart
2) Можно для всего тулбара использовать androidx.appcompat.widget.Toolbar
3) Можно оставить как сейчас, но задавать вертикальный gravity (или layout_gravity, постоянно забываю кто из них кто)

faritowich commented on app/src/main/res/layout/activity_settings.xml:
```diff
@@ -0,0 +1,121 @@
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center"
+            android:layout_marginStart="@dimen/settings_text_marginStart"
```
Лучше не прописывать отступы в каждом элементе, а задать общий отступ в родительском лейауте

KiMate34 committed:
Remade set on click listener on buttonBackMain
Changed dimensions
Changed font textSize to main_interface_textSize
Delete extra FrameLayout in activity_settings.xml
Take it out iconPadding and iconSize in activity_settings.xml
Delete extra dimensions TextView in activity_settings.xml
Changed TextView in activity_settings.xml

KiMate34 committed:
Added dark theme for all activities
Added functionality for activities settings
Added functionality for search activity

faritowich commented on app/src/main/java/com/example/playlistmaker/SettingsActivity.kt:
```diff
@@ -0,0 +1,60 @@
+                putExtra(Intent.EXTRA_TEXT, getString(R.string.support_message))
+            }
+            startActivity(sendSupportIntent)
+        }
```
Можно упростить до
```
            Intent(Intent.ACTION_SENDTO).apply {
                data = Uri.parse("mailto:")
                putExtra(Intent.EXTRA_EMAIL, arrayOf(getString(R.string.support_address)))
                putExtra(Intent.EXTRA_SUBJECT, getString(R.string.support_subject))
                putExtra(Intent.EXTRA_TEXT, getString(R.string.support_message))
                startActivity(this)
            }
```

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -0,0 +1,68 @@
+import android.widget.Button
+import android.widget.EditText
+import android.widget.ImageView
+import android.widget.LinearLayout
```
Неиспользуемые импорты лучше удалять

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -0,0 +1,68 @@
+        inputSearchText.addTextChangedListener(simpleTextWatcher)
+
+    }
+    companion object {
```
companion object принято объявлять в конце класса (реже в начале), но не посреди других методов класса

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -0,0 +1,68 @@
+
+    }
+    companion object {
+        var SEARCH_TEXT = ""
```
В companion object принято выносить константы, а переменные для конкретного экземпляра активити надо объявлять в самом классе.
Это связано с тем, что SEARCH_TEXT логически связано с конкретным экземпляром активити. А экземпляров класса может быть сколько угодно, а сейчас SEARCH_TEXT привязан не к конкретному экземпляру класса, а к классу в целом. Константа же обычно не привязана к конкретному экземпляру, так что их выносят в companion object - так константа будет одна для всех экземпляров класса, даже если их будет сто. А переменная, хранящая введенный текст, должна быть привязана к конкретному экземпляру активити.

То, что постоянно все зависимости от экземпляра класса - в константу в companion object, иначе - просто переменная.

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -0,0 +1,68 @@
+
+    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
+        super.onRestoreInstanceState(savedInstanceState)
+        savedInstanceState.getString(SEARCH_TEXT)
```
Что делается с полученной стрингой в этой строке кода? Строка просто получается и нигде не используется

faritowich commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,89 @@
+        <item name="android:layout_width">wrap_content</item>
+        <item name="android:layout_height">wrap_content</item>
+        <item name="android:layout_gravity">end</item>
+        <item name="android:layout_marginTop">@dimen/settings_imageView_marginTop</item>
```
Лучше внешние отступы, высоту и ширину, веса, гравити не выносить в стиль. Если на экране будет много элементов, будет сложнее ориентироваться с их взаимным расположением, когда отвечающие за это атрибуты разнесены по разным файлам - что-то может быть в лейауте, что-то в стиле. В стилях обычно содержат атрибуты, отвечающие за внешний вид вьюхи и вообще только за то, что находится внутри нее - цвета, шрифты, фоны и т.д. Бывают исключения, когда, например, мы точно знаем, что у кнопки всегда должна строго определенная высота и ширина - тогда да, имеет смысл выносить

KiMate34 committed:
On SearchActivity:
Change job clearButton
Change onSaveInstanceState, onRestoreInstanceState
On activity_search:
Change view EditText

KiMate34 committed:
On activity_search:
Change view EditText
Change view ImageView

KiMate34 committed:
Add RecyclerView in activity_search.xml
Create classes: Track, TrackAdapter, TrackViewHolder
